File: python/src/azoth/eos/reference/hydrate_formation_temperature.py
"""``eos.hydrate_formation_temperature``: the Python reference.

The mirror of ``crates/azoth-eos/src/hydrate_formation_temperature.rs``:

```text
solve  f_w^hydrate(T, P, fugacities) / f_w^fluid(T, P) - 1 = 0   for T
```

Each trial is a full flash of the fluid, because the guests' fugacities come from the gas it
produces - what NeqSim's ``setFug`` copies in before every occupancy evaluation - and the
hydrate's water fugacity is then built from those by :mod:`azoth.eos.reference._hydrate`.

The answer is settled by **water**: the hydrate and the fluid meet where water's fugacity is
the same in both. The structure is not an input but an output of the same comparison, because
the lower of the two coefficients is the stable hydrate.
"""

from __future__ import annotations

import math

from azoth.core.errors import InvalidInputError, SolverNotConvergedError
from azoth.core.range import apply_checks, checks_for
from azoth.core.result import (
    HydrateFormationTemperatureResult,
    HydrateStructure,
    PtFlashResult,
)
from azoth.core.units import Q, from_si, input_to_si
from azoth.core.warnings import Warning
from azoth.eos.reference import _hydrate
from azoth.eos.reference._mixture_state import phase_state, reduced_parameters
from azoth.eos.reference.pt_flash import pt_flash

MODEL_ID = "eos.hydrate_formation_temperature"


def _reference_water_fugacity(eos: str, t: float, p: float) -> float:
    """The fugacity of pure water at a state on the cubic the fluid runs, in Pa.

    NeqSim's ``refPhase`` is a one-component phase of the *host's* class - an SRK phase for
    an SRK fluid, not ``ComponentWater`` - and this builds the same thing: the same cubic,
    water alone, on the liquid root.
    """
    from azoth.eos import components as databank

    water, _ = databank.mixture_of(["water"], eos=eos)
    reduced = reduced_parameters(water, t, p)
    state = phase_state(reduced, water.kij, [1.0], liquid=True)
    return math.exp(state.ln_phi[0]) * p


def _guest_fugacities(flash: PtFlashResult, z: list[float], p: float) -> list[float]:
    """The guests' fugacities at a flash, in Pa, from the phase ``setFug`` reads.

    The **vapour** phase's, which is what NeqSim's ``setFug`` copies in. A feed the flash
    finds single-phase has no vapour to read, so its own composition is its phase
    composition, with the coefficients of whichever phase the flash says it is.
    """
    two_phase = getattr(flash, "phase", "") == "two_phase"
    composition = list(flash.y) if two_phase else list(z)
    coefficients = flash.ln_phi_liquid if flash.phase == "all_liquid" else flash.ln_phi_vapour
    return [
        fraction * math.exp(coefficient) * p
        for fraction, coefficient in zip(composition, coefficients, strict=True)
    ]


def hydrate_formation_temperature(
    components: list[str], P: Q, z: list[float], eos: str = "srk"
) -> HydrateFormationTemperatureResult:
    """The temperature at which a fluid's hydrate appears, at a pressure.

    **The components cross by name**, as they do for the Fürst phase, because the hydrate's
    guest tables are keyed by name and a :class:`~azoth.eos.mixture.Component` carries none.

    Args:
        components: the substances, by name.
        P: absolute pressure; the temperature is what is solved for.
        z: the overall mole fractions. Checked rather than renormalised.
        eos: the cubic the fluid runs, which is also the one the hydrate's reference water
            phase is built from.

    Raises:
        InvalidInputError: if the fluid has no water, or nothing in it occupies a cage.
        OutOfRangeError: if ``P`` is not positive or a trial's cavity sum has no value.
        SolverNotConvergedError: if the scan finds no sign change or the bisection hits its
            cap.
    """
    from azoth import _models_gen
    from azoth.eos import components as databank

    spec = _models_gen.model(MODEL_ID)
    checks = checks_for(spec)
    warnings: list[Warning] = []
    p_si = input_to_si(spec, "P", P)
    apply_checks(checks.on_input, {"P": p_si}.get, warnings)

    hydration = _hydrate.hydration_for(components)
    if hydration.water_index is None:
        raise InvalidInputError("components", "the hydration carries no water index")
    water_index = hydration.water_index
    mixture, _ = databank.mixture_of(components, eos=eos)

    algorithm = spec["algorithm"]
    bracket = algorithm["bracket"]
    tolerance = float(algorithm["tolerance"])
    steps = max(int(bracket["steps"]), 2)

    def residual(t: float) -> float:
        flash = pt_flash(mixture, from_si(t, "K"), from_si(p_si, "Pa"), z)
        fugacities = _guest_fugacities(flash, z, p_si)
        reference = _reference_water_fugacity(eos, t, p_si)
        _structure, coefficient = _hydrate.stable_structure(
            hydration.guests, fugacities, t, p_si, reference
        )
        return coefficient * p_si / fugacities[water_index] - 1.0

    iterations = 0
    previous: tuple[float, float] | None = None
    found: tuple[float, float] | None = None
    lower = float(bracket["lower"])
    upper = float(bracket["upper"])
    for step in range(steps):
        t = lower + (upper - lower) * (step / (steps - 1))
        value = residual(t)
        iterations += 1
        if previous is not None and previous[1] * value <= 0.0:
            found = (previous[0], t)
            break
        previous = (t, value)

    if found is None:
        raise SolverNotConvergedError(
            iterations=iterations,
            residual=math.nan if previous is None else previous[1],
            tolerance=tolerance,
        )

    low, high = found
    low_value = residual(low)
    iterations += 1
    for _ in range(int(algorithm["max_iterations"])):
        mid = 0.5 * (low + high)
        value = residual(mid)
        iterations += 1
        if abs(value) < tolerance or high - low < 1.0e-9:
            flash = pt_flash(mixture, from_si(mid, "K"), from_si(p_si, "Pa"), z)
            fugacities = _guest_fugacities(flash, z, p_si)
            reference = _reference_water_fugacity(eos, mid, p_si)
            structure, _coefficient = _hydrate.stable_structure(
                hydration.guests, fugacities, mid, p_si, reference
            )
            return HydrateFormationTemperatureResult(
                temperature=from_si(mid, "K"),
                structure=(
                    HydrateStructure.STRUCTURE_I
                    if structure == 0
                    else HydrateStructure.STRUCTURE_II
                ),
                iterations=iterations,
                residual=value,
                warnings=tuple(warnings),
            )
        if low_value * value <= 0.0:
            high = mid
        else:
            low, low_value = mid, value

    raise SolverNotConvergedError(
        iterations=iterations,
        residual=0.5 * (low + high),
        tolerance=tolerance,
    )


__all__ = ["hydrate_formation_temperature"]

File: crates/azoth-eos/src/hydrate_formation_temperature.rs
//! `eos.hydrate_formation_temperature` - the temperature at which a fluid's hydrate appears.
//!
//! ```text
//! solve  f_w^hydrate(T, P, fugacities) / f_w^fluid(T, P) - 1 = 0   for T
//! ```
//!
//! NeqSim's `HydrateFormationTemperatureFlash`. Each trial is a full flash of the fluid,
//! because the guests' fugacities come from the gas it produces - which is what NeqSim's
//! `setFug` copies in before every occupancy evaluation - and then the hydrate's water
//! fugacity is built from those by [`crate::hydrate`].
//!
//! # What settles the answer
//!
//! Water. The hydrate and the fluid meet where water's fugacity is the same in both, and the
//! two structures are evaluated at every trial because the **lower** coefficient is the stable
//! hydrate - so the structure is an output of the same comparison the temperature is.
//!
//! # The reference, and why the fluid's equation is an input
//!
//! The coefficient's third term is `ln(f_w^ref/f_w^fluid)`, and `f_w^ref` is the fugacity of
//! pure water on **the same cubic the fluid runs** - NeqSim builds it from a one-component
//! phase of the host's own class. It is computed here from a water-only mixture rather than
//! taken, so the model needs to know the equation; it reads it from the mixture itself.

use azoth_core::units::{Pressure, kelvins};
use azoth_core::{AzothError, Result, apply_checks};

use crate::databank;
use crate::hydrate;
use crate::mixture::{Mixture, RootSide};
use crate::model_gen;
use crate::pt_flash::pt_flash;
use crate::results::{HydrateFormationTemperatureResult, HydrateStructure, PtFlashResult};

/// The fugacity of pure water at a state on the cubic the fluid runs, in Pa.
///
/// NeqSim's `refPhase` is a one-component phase of the *host's* class - an SRK phase for an
/// SRK fluid, not `ComponentWater` - and this builds the same thing: the same cubic, water
/// alone, on the liquid root.
///
/// # Errors
/// * [`AzothError`] from the water-only mixture's own state.
fn reference_water_fugacity(mixture: &Mixture, t: f64, p: f64) -> Result<f64> {
    let (water, _) = databank::mixture_of(&["water"], mixture.cubic(), None)?;
    let reduced = water.reduced_parameters(kelvins(t), azoth_core::units::pascals(p))?;
    let state = water.phase_state(&reduced, &[1.0], RootSide::Liquid)?;
    Ok(state.ln_phi[0].exp() * p)
}

/// The guests' fugacities at a flash of the fluid, in Pa.
///
/// From the **vapour** phase's composition and coefficients, which is the phase NeqSim's
/// `setFug` reads. A single-phase feed has no vapour to read, so the feed's own composition is
/// its phase composition - and the fugacity coefficients come from whichever phase the flash
/// says it is.
fn guest_fugacities(flash: &PtFlashResult, z: &[f64], p: f64) -> Vec<f64> {
    let composition = match flash.phase {
        crate::results::Phase::TwoPhase => flash.y.clone(),
        _ => z.to_vec(),
    };
    let ln_phi = match flash.phase {
        crate::results::Phase::AllLiquid => &flash.ln_phi_liquid,
        _ => &flash.ln_phi_vapour,
    };
    composition
        .iter()
        .zip(ln_phi)
        .map(|(fraction, coefficient)| fraction * coefficient.exp() * p)
        .collect()
}

/// The temperature at which a fluid's hydrate appears, at a pressure.
///
/// # Errors
/// * [`AzothError::InvalidInput`] if the mixture was not resolved for a hydrate - build it
///   with [`hydrate::hydrate_mixture_of`], which is what attaches the guests' tables.
/// * [`AzothError::OutOfRange`] if `P` is not positive, or if a trial temperature's cavity sum
///   has no value.
/// * [`AzothError::SolverNotConverged`] if the scan finds no sign change or the bisection hits
///   its cap.
pub fn hydrate_formation_temperature(
    mixture: &Mixture,
    p: Pressure,
    z: &[f64],
) -> Result<HydrateFormationTemperatureResult> {
    let spec = &model_gen::HYDRATE_FORMATION_TEMPERATURE_SPEC;
    let mut warnings = Vec::new();
    apply_checks(
        spec.input_checks(),
        |quantity| (quantity == "P").then_some(p.value),
        &mut warnings,
    )?;

    let hydration = mixture.hydration().ok_or_else(|| {
        AzothError::invalid_input(
            "components",
            "this mixture was not resolved for a hydrate; `hydrate_mixture_of` is what \
             attaches the guests' tables"
                .to_string(),
        )
    })?;
    let water_index = hydration
        .water_index
        .expect("the resolver refuses no water");

    let algorithm = crate::algorithm_of(spec)?;
    let bracket = algorithm.bracket.ok_or_else(|| {
        AzothError::invalid_input(
            "algorithm",
            "this procedure needs a `[algorithm.bracket]` to scan for the sign change, and \
             the spec declares none"
                .to_string(),
        )
    })?;

    let residual = |t: f64| -> Result<f64> {
        let flash = pt_flash(mixture, kelvins(t), p, z)?;
        let fugacities = guest_fugacities(&flash, z, p.value);
        let reference = reference_water_fugacity(mixture, t, p.value)?;
        let (_structure, coefficient) = hydrate::stable_structure(
            &hydration.guests,
            &fugacities,
            hydration.model,
            t,
            p.value,
            reference,
        )?;
        Ok(coefficient * p.value / fugacities[water_index] - 1.0)
    };

    // The scan, then the bisection: a hydrate's formation temperature is monotone in pressure
    // at fixed composition, and a linear scan is what the spec declares.
    let steps = bracket.steps.max(2);
    let mut iterations = 0;
    let mut lower: Option<(f64, f64)> = None;
    let mut previous: Option<(f64, f64)> = None;
    let mut bracket_found = None;
    for step in 0..steps {
        let fraction = step as f64 / (steps - 1) as f64;
        let t = bracket.lower + (bracket.upper - bracket.lower) * fraction;
        let value = residual(t)?;
        iterations += 1;
        if let Some((previous_t, previous_value)) = previous
            && previous_value * value <= 0.0
        {
            bracket_found = Some((previous_t, t));
            break;
        }
        previous = Some((t, value));
        lower = Some((t, value));
    }
    let (mut lo, mut hi) = bracket_found.ok_or(AzothError::SolverNotConverged {
        iterations,
        residual: lower.map(|(_, value)| value).unwrap_or(f64::NAN),
        tolerance: algorithm.tolerance,
    })?;
    let mut lo_value = residual(lo)?;
    iterations += 1;

    for _ in 0..algorithm.max_iterations {
        let mid = 0.5 * (lo + hi);
        let value = residual(mid)?;
        iterations += 1;
        if value.abs() < algorithm.tolerance || hi - lo < 1.0e-9 {
            return finish(mixture, p, z, hydration, mid, value, iterations, warnings);
        }
        if lo_value * value <= 0.0 {
            hi = mid;
        } else {
            lo = mid;
            lo_value = value;
        }
    }

    Err(AzothError::SolverNotConverged {
        iterations,
        residual: 0.5 * (lo + hi),
        tolerance: algorithm.tolerance,
    })
}

/// The answer, with the structure the same comparison chose.
#[allow(clippy::too_many_arguments)]
fn finish(
    mixture: &Mixture,
    p: Pressure,
    z: &[f64],
    hydration: &hydrate::Hydration,
    temperature: f64,
    residual: f64,
    iterations: u32,
    warnings: Vec<azoth_core::Warning>,
) -> Result<HydrateFormationTemperatureResult> {
    let flash = pt_flash(mixture, kelvins(temperature), p, z)?;
    let fugacities = guest_fugacities(&flash, z, p.value);
    let reference = reference_water_fugacity(mixture, temperature, p.value)?;
    let (structure, _coefficient) = hydrate::stable_structure(
        &hydration.guests,
        &fugacities,
        hydration.model,
        temperature,
        p.value,
        reference,
    )?;

    Ok(HydrateFormationTemperatureResult {
        temperature: kelvins(temperature),
        structure: if structure == 0 {
            HydrateStructure::StructureI
        } else {
            HydrateStructure::StructureIi
        },
        iterations,
        residual,
        warnings,
    })
}
